### ISSUE TITLE
Provide examples in API documentation

### ISSUE THREAD

**Issue by malcolmr** [nodatime/nodatime#819]
I've been thinking that it would be good to include examples in the API documentation to illustrate common patterns, or just basic usage, for common types. (Similar to http://nodatime.org/userguide/recipes, I guess.)

Looking at the results from Analytics, the following are the most-referenced pages (by total time on page, >10 unique page views over the 9-day period), across all API versions:

```
/NodaTime.LocalDateTime.html
/NodaTime.DateTimeZone.html
/NodaTime.Instant.html
/NodaTime.TimeZones.TzdbDateTimeZoneSource.html
/NodaTime.ZonedDateTime.Comparer.html
/NodaTime.ZonedDateTime.html
/NodaTime.LocalDate.html
/NodaTime.Interval.html
```
```
/NodaTime.CalendarSystem.html
/NodaTime.Duration.html
/NodaTime.Offset.html
/NodaTime.DateTimeZoneProviders.html
/NodaTime.OffsetDateTime.html
/NodaTime.Text.LocalDatePattern.html
/NodaTime.DateInterval.html
/NodaTime.IClock.html
/NodaTime.Testing.FakeClock.html
/NodaTime.Period.html
/NodaTime.SystemClock.html
/NodaTime.LocalTime.html
```

The first group represents 50% of time spent on API documentation; the second, a further 30%.

@jskeet I assume there's no better way of doing this than adding type documentation with e.g.

```csharp
/// </summary>
/// <example>
/// This sample shows how to Noda Time:
/// <code>
/// ...
/// </code>
/// Here's the output:
/// <code>...</code>
/// </example>
```

... and then extracting and running the examples by hand?

**Comment by jskeet** [comment]
As it happens, there *is* a better way :) We can use the same approach as we do in https://github.com/GoogleCloudPlatform/google-cloud-dotnet
- we have a separate Snippets project, written as tests, then we extract this snippets and add them into the docs as part of the docfx build.

Will need to work out how to separate out the snippet build code though...

**Comment by malcolmr** [comment]
Alright, that sounds sensible. No reason we can't start a snippets project now (or perhaps we should use `NodaTime.Demo`? That's basically what it is?) and work on the documentation integration later?

**Comment by jskeet** [comment]
Yes, let's do it in NodaTime.Demo - that would work fine.

Given that the target is only NodaTime, the integration can probably be simpler - and here may be a better place to experiment with Roslyn integration more closely. It would be *really* nice to be able to autodetect the required `using` directives and include those...

**Comment by jskeet** [comment]
Doing cunning stuff with Roslyn is probably blocked on https://github.com/dotnet/roslyn/issues/17968

**Comment by jskeet** [comment]
Although I could probably create a separate "old school" msbuild project that referred to the nuget package. That would at least be enough to play with for now...

**Comment by jskeet** [comment]
Resolved that issue, so it's all looking feasible again...

**Comment by malcolmr** [comment]
(#821 implements the start of this.)

**Comment by jskeet** [comment]
Closing this in favour of #918.
